content/grants/persistent-memory.md

```diff
@@ -0,0 +1,158 @@
+
+Thus, it should suffice to maintain a single, global writer lock for an arena.
+Contention for this lock is not likely to be high, and this will eliminate a great deal of complexity which would obtain if page dirtying and commits required more granular synchronization.
+
+## Tuning
+
+Commit frequency of the Arvo snapshot arena is a potentially tunable value, with performance implications.
+Committing more often incurs more disk IO, saving nouns that will potentially soon be discarded (Ames queues, for instance). However, under memory pressure, it makes paging more performant, as the proportion of dirty pages which must be *written* to disk to evict is smaller. Already-committed pages can simply be evicted from memory and re-read from disk when a read fault occurs. Committing less often requires writes as part of eviction, but reduces disk IO. It is desirable as long as memory pressure as low, and highly desirable for block storage where write-wearing is a concern.
+
+
+# Evaluation
+An important part of this work is to objectively evaluate the proposed approach. This will require the implementation of at least one alternate approach. The primary alternative is to map pages `MAP_PRIVATE` and simply track dirty pages and set `PROT_WRITE` in the fault handler. Then dirty pages could be more lazily written to disk, using `write()`, and committing using `fsync()`. The main reason this tradeoff is not expected to perform well is that it requires the underlying OS to have swap space configured and available if a dirty page must be evicted from the page cache. Another way of stating this is that the size of our dirty page set is limited by available RAM and swap space.
+
+However, this approach is used both by the current Vere as well as by LMDB for writing dirtied pages. It should therefore be implemented for New Mars at least by way of comparison.
+
+# Milestones & Compensation
+
+## Milestone 1: Basic copy-on-write arena
+
+Implement a copy-on-write arena by using `write()` to copy pages to new regions of the backing file.
+This should include at least single-arena snapshot support.
+
+Compensation: 1 star
+
+## Milestone 2: Kernel copy-on-write (`MAP_PRIVATE`) + lazy write arena
+
+Implement a copy-on-write arena by mapping pages `MAP_PRIVATE` and lazily writing dirty pages to the backing file.
+
+Compensation: 1 star
+
+## Milestone 3: Memory allocation and separate event arena
+
+Implement a memory allocator on top of this arena. This includes the ability to store metadata such as roots which will be committed with the snapshot, and reference counting or tracing to free nouns which are not referenced from the root.
+
+Implement a separate (may be interleaved at the page level in memory/on disk) arena for event logs, which is bump-allocated and not copy-on-write but rather does not reallocate in committed pages.
+
+Compensation: 2 stars
```

Comment:
I'm more inclined to having a milestone for making the "basic copy-on-write arena" less basic (adding a b+tree, maybe optimizing copy destinations for adjacency so as to minimize page table entries, or anything else that comes up). I'm not sure how much the details of the actual allocator/arena will be tied to and require coordination with the larger new-mars project.

Reply:
I think something like a B tree or B+ tree for the index is a necessary part of the initial implementation actually. But it's also not hard and can be implemented in that few-weeks timeframe.

Reply:
Ok, sounds good. I wasn't sure just how minimal we were talking.